numpyro/distributions/kl.py

```diff
@@ -0,0 +1,180 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+# The implementation follows the design in PyTorch: torch.distributions.kl.py
```

Comment:
@fehiepsi it might be simpler to use Matt Rocklin's [multipledispatch](https://github.com/mrocklin/multipledispatch) library, which we use in Funsor. The original motivation for adding this pairwise KL stuff in PyTorch is that PyTorch didn't want to add multipledispatch as a dependency. But since NumPyro already depends on Funsor, you won't be adding a dependency.
```py
from multipledispatch import dispatch

@dispatch(Distribution, ExpandedDistribution)
def kl_divergence(p, q):
    kl = kl_divergence(p, q.base_dist)
    shape = lax.broadcast_shapes(p.batch_shape, q.batch_shape)
    return jnp.broadcast_to(kl, shape)

...
```

Reply:
Yeah, it is a good idea. But `funsor` is currently an optional dependency of numpyro, so I would like to defer it to a future PR. I'll make an issue to track this.